script/prerender/manage.ts

```diff
@@ -0,0 +1,397 @@
+        break;
+      case 'DeadLetterQueueUrl':
+        deadLetterQueueUrl = outputValue;
+        break;
+    }
+  }
+
+  if (!workQueueUrl) {
+    throw new Error(`${WORKERS_STACK_NAME} stack did not have a WorkQueueUrl output`);
+  }
+
+  if (!deadLetterQueueUrl) {
+    throw new Error(`${WORKERS_STACK_NAME} stack did not have a DeadLetterQueueUrl output`);
+  }
+
+  console.log('Waiting for all jobs to be queued');
+
+  // Wait for all book pages to be queued
+  const books = await allPagesQueuedPromise;
+
+  numBooks = books.length;
+
+  console.log(`All ${numPages} page prerendering jobs and all ${numBooks} sitemap jobs queued`);
+
+  // console.log('Queuing sitemap index job');
+
+  await sqsClient.send(new SendMessageCommand({
+    MessageBody: JSON.stringify({ payload: { books }, type: 'sitemapIndex' }),
+    QueueUrl: workQueueUrl,
+  }));
+
+  console.log('1 sitemap index job queued');
+
+  const numJobs = numPages + numBooks + 1;
+
+  // Ensure the work queue is empty
+
+  console.log('Waiting 1 minute for the work queue attributes to stabilize');
+
+  // First wait 1 minute after sending the last message for the queue attributes to stabilize
+  // This is required according to SQS docs
+  await new Promise((resolve) => setTimeout(resolve, 60000));
+
+  console.log('Waiting for the work queue to be empty');
+
+  // Now check that all the NumberOfMessages attributes are 0
+  const getQueueAttributesCommand = new GetQueueAttributesCommand({
+    AttributeNames: [
+      'ApproximateNumberOfMessages',
+      'ApproximateNumberOfMessagesDelayed',
+      'ApproximateNumberOfMessagesNotVisible',
+    ],
+    QueueUrl: workQueueUrl,
+  });
+  while (true) {
+    const getQueueAttributesResult = await sqsClient.send(getQueueAttributesCommand);
+    const attributes = assertDefined(
+      getQueueAttributesResult.Attributes,
+      '[SQS] [GetQueueAttributes] Unexpected response: missing Attributes key'
+    );
+
+    const numMessages = parseInt(attributes.ApproximateNumberOfMessages, 10) +
+      parseInt(attributes.ApproximateNumberOfMessagesDelayed, 10) +
+      parseInt(attributes.ApproximateNumberOfMessagesNotVisible, 10);
+
+    if (numMessages === 0) { break; }
+
+    console.log(`${numMessages}/${numJobs} prerendering jobs remaining`);
+
+    if (new Date() > timeoutDate) {
+      throw new Error(`Not all prerendering jobs finished within ${
+        PRERENDER_TIMEOUT_SECONDS} seconds of stack creation.`);
+    }
+
+    await new Promise((resolve) => setTimeout(resolve, 10000));
+  }
+
+  console.log(`All ${numJobs} prerendering jobs finished`);
+
+  console.log('Ensuring that the dead letter queue is empty');
+
+  // Ensure that the dead letter queue is also empty
+  // Since we are the only consumer of the dead letter queue, we use long polling to check
+  // that it is empty rather than waiting another minute and checking the queue attributes
+  const receiveDLQMessageResult = await sqsClient.send(new ReceiveMessageCommand({
```

Comment:
this removes the messages from the queue right? wouldn't that prevent us from being able to redrive them?

Reply:
Normally this would make the messages invisible, but the DLQ has the VisibilityTimeout set to 0 which may or may not mean messages can never be successfully deleted from it.
However, the DLQ is deleted when the workers stack is deleted at the end of the build, even if it failed, which happens right after the manager detects any messages in the DLQ.
If you're thinking about wanting to retry failed builds (not from the beginning), then we'd maybe have to split the DLQ into a separate stack.

Reply:
it may become convenient to have this script try to redrive the dlq once or twice before exiting if the build becomes flaky, not necessarily right now but it would be nice if we didn't have to redo this part of the logic if we decide to do that

Reply:
I think that is handled in the normal queue. If the worker doesn't delete the message it'll be retried by another worker when the visibility timeout expires. The DLQ only comes into play when the message has already failed some (configurable) number of times. To retry more, we can just increase that number instead.